wallet/src/restore.rs

```diff
@@ -103,85 +112,158 @@ pub fn utxos_batch_block(config: &WalletConfig, start_height: u64, end_height:u6
 	}
 }
 
-fn find_utxos_with_key(config:&WalletConfig, keychain: &Keychain, 
-	switch_commit_cache : &Vec<[u8;SWITCH_COMMIT_HASH_SIZE]>,
-	block_outputs:api::BlockOutputs, key_iterations: &mut usize, padding: &mut usize) 
-	-> Vec<(pedersen::Commitment, Identifier, u32, u64, u64, bool) > {
-	//let key_id = keychain.clone().root_key_id();
-	let mut wallet_outputs: Vec<(pedersen::Commitment, Identifier, u32, u64, u64, bool)> = Vec::new();
+fn find_utxos_with_key(
+	config: &WalletConfig,
+	keychain: &Keychain,
+	switch_commit_cache: &Vec<[u8; SWITCH_COMMIT_HASH_SIZE]>,
+	block_outputs: api::BlockOutputs,
+	key_iterations: &mut usize,
+	padding: &mut usize,
+) -> Vec<(pedersen::Commitment, Identifier, u32, u64, u64, bool)> {
+	let mut wallet_outputs: Vec<(pedersen::Commitment, Identifier, u32, u64, u64, bool)> =
+		Vec::new();
+
+	info!(
+		LOGGER,
+		"Scanning block {}, {} outputs, over {} key derivations",
+		block_outputs.header.height,
+		block_outputs.outputs.len(),
+		*key_iterations,
+	);
 
-	info!(LOGGER, "Scanning block {} over {} key derivation possibilities.", block_outputs.header.height, *key_iterations);
 	for output in block_outputs.outputs {
 		for i in 0..*key_iterations {
-			if switch_commit_cache[i as usize]==output.switch_commit_hash {
-				info!(LOGGER, "Output found: {:?}, key_index: {:?}", output.switch_commit_hash,i);
-				//add it to result set here
+			if switch_commit_cache[i as usize] == output.switch_commit_hash {
+				info!(
+					LOGGER,
+					"Output found: {:?}, key_index: {:?}",
+					output.switch_commit_hash,
+					i,
+				);
+
+				// add it to result set here
 				let commit_id = from_hex(output.commit.clone()).unwrap();
 				let key_id = keychain.derive_key_id(i as u32).unwrap();
-				let (amount, is_coinbase) = retrieve_amount_and_coinbase_status(config,
-					keychain, key_id.clone(), &output.commit);
-				info!(LOGGER, "Amount: {}", amount);
-				let commit = keychain.commit_with_key_index(BigEndian::read_u64(&commit_id), i as u32).unwrap();
-				wallet_outputs.push((commit, key_id.clone(), i as u32, amount, output.height, is_coinbase));
-				//probably don't have to look for indexes greater than this now
-				*key_iterations=i+*padding;
-				if *key_iterations > switch_commit_cache.len() {
-					*key_iterations = switch_commit_cache.len();
+				let res = retrieve_amount_and_coinbase_status(
+					config,
+					keychain,
+					key_id.clone(),
+					&output.commit,
+				);
+
+				if let Ok((amount, is_coinbase)) = res {
```

Comment:
This is the additional error handling.
We may want to investigate this further - but it seems to be affected by some early blocks where we cannot safely recover the amount.